From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcrt: Implement _findfirst32i64 and _findnext32i64 with their wide forms.

_findfirst32i64, _findnext32i64, _wfindfirst32i64 and _wfindnext32i64 (32-bit
times, 64-bit sizes) were spec stubs in msvcr80 .. msvcr120 and ucrtbase
(with its _o_ forms); msvcr120_app forwards.
---
 dlls/msvcr100/msvcr100.spec         |   8 +--
 dlls/msvcr110/msvcr110.spec         |   8 +--
 dlls/msvcr120/msvcr120.spec         |   8 +--
 dlls/msvcr120_app/msvcr120_app.spec |   8 +--
 dlls/msvcr80/msvcr80.spec           |   8 +--
 dlls/msvcr90/msvcr90.spec           |   8 +--
 dlls/msvcrt/dir.c                   | 113 ++++++++++++++++++++++++++++++++++++
 dlls/ucrtbase/ucrtbase.spec         |  16 ++---
 8 files changed, 145 insertions(+), 32 deletions(-)

diff --git a/dlls/msvcr100/msvcr100.spec b/dlls/msvcr100/msvcr100.spec
index eaf3324..a9dd06e 100644
--- a/dlls/msvcr100/msvcr100.spec
+++ b/dlls/msvcr100/msvcr100.spec
@@ -982,11 +982,11 @@
 @ cdecl _fileno(ptr)
 @ cdecl _findclose(long)
 @ cdecl _findfirst32(str ptr)
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr)
 @ cdecl _findfirst64(str ptr)
 @ cdecl _findfirst64i32(str ptr)
 @ cdecl _findnext32(long ptr)
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr)
 @ cdecl _findnext64(long ptr)
 @ cdecl _findnext64i32(long ptr)
 @ cdecl _finite(double)
@@ -1715,11 +1715,11 @@
 @ cdecl _wexecvpe(wstr ptr ptr)
 @ cdecl _wfdopen(long wstr)
 @ cdecl _wfindfirst32(wstr ptr)
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr)
 @ cdecl _wfindfirst64(wstr ptr)
 @ cdecl _wfindfirst64i32(wstr ptr)
 @ cdecl _wfindnext32(long ptr)
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr)
 @ cdecl _wfindnext64(long ptr)
 @ cdecl _wfindnext64i32(long ptr)
 @ cdecl _wfopen(wstr wstr)
diff --git a/dlls/msvcr110/msvcr110.spec b/dlls/msvcr110/msvcr110.spec
index 52abeca..3aa39e3 100644
--- a/dlls/msvcr110/msvcr110.spec
+++ b/dlls/msvcr110/msvcr110.spec
@@ -1158,11 +1158,11 @@
 @ cdecl _fileno(ptr)
 @ cdecl _findclose(long)
 @ cdecl _findfirst32(str ptr)
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr)
 @ cdecl _findfirst64(str ptr)
 @ cdecl _findfirst64i32(str ptr)
 @ cdecl _findnext32(long ptr)
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr)
 @ cdecl _findnext64(long ptr)
 @ cdecl _findnext64i32(long ptr)
 @ cdecl _finite(double)
@@ -1902,11 +1902,11 @@
 @ cdecl _wexecvpe(wstr ptr ptr)
 @ cdecl _wfdopen(long wstr)
 @ cdecl _wfindfirst32(wstr ptr)
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr)
 @ cdecl _wfindfirst64(wstr ptr)
 @ cdecl _wfindfirst64i32(wstr ptr)
 @ cdecl _wfindnext32(long ptr)
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr)
 @ cdecl _wfindnext64(long ptr)
 @ cdecl _wfindnext64i32(long ptr)
 @ cdecl _wfopen(wstr wstr)
diff --git a/dlls/msvcr120/msvcr120.spec b/dlls/msvcr120/msvcr120.spec
index b3da77c..c3cee63 100644
--- a/dlls/msvcr120/msvcr120.spec
+++ b/dlls/msvcr120/msvcr120.spec
@@ -1159,11 +1159,11 @@
 @ cdecl _fileno(ptr)
 @ cdecl _findclose(long)
 @ cdecl _findfirst32(str ptr)
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr)
 @ cdecl _findfirst64(str ptr)
 @ cdecl _findfirst64i32(str ptr)
 @ cdecl _findnext32(long ptr)
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr)
 @ cdecl _findnext64(long ptr)
 @ cdecl _findnext64i32(long ptr)
 @ cdecl _finite(double)
@@ -1927,11 +1927,11 @@
 @ cdecl _wexecvpe(wstr ptr ptr)
 @ cdecl _wfdopen(long wstr)
 @ cdecl _wfindfirst32(wstr ptr)
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr)
 @ cdecl _wfindfirst64(wstr ptr)
 @ cdecl _wfindfirst64i32(wstr ptr)
 @ cdecl _wfindnext32(long ptr)
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr)
 @ cdecl _wfindnext64(long ptr)
 @ cdecl _wfindnext64i32(long ptr)
 @ cdecl _wfopen(wstr wstr)
diff --git a/dlls/msvcr120_app/msvcr120_app.spec b/dlls/msvcr120_app/msvcr120_app.spec
index ff88090..615b4cb 100644
--- a/dlls/msvcr120_app/msvcr120_app.spec
+++ b/dlls/msvcr120_app/msvcr120_app.spec
@@ -1088,11 +1088,11 @@
 @ cdecl _fileno(ptr) msvcr120._fileno
 @ cdecl _findclose(long) msvcr120._findclose
 @ cdecl _findfirst32(str ptr) msvcr120._findfirst32
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr) msvcr120._findfirst32i64
 @ cdecl _findfirst64(str ptr) msvcr120._findfirst64
 @ cdecl _findfirst64i32(str ptr) msvcr120._findfirst64i32
 @ cdecl _findnext32(long ptr) msvcr120._findnext32
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr) msvcr120._findnext32i64
 @ cdecl _findnext64(long ptr) msvcr120._findnext64
 @ cdecl _findnext64i32(long ptr) msvcr120._findnext64i32
 @ cdecl _finite(double) msvcr120._finite
@@ -1612,11 +1612,11 @@
 @ extern _wenviron msvcr120._wenviron
 @ cdecl _wfdopen(long wstr) msvcr120._wfdopen
 @ cdecl _wfindfirst32(wstr ptr) msvcr120._wfindfirst32
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr) msvcr120._wfindfirst32i64
 @ cdecl _wfindfirst64(wstr ptr) msvcr120._wfindfirst64
 @ cdecl _wfindfirst64i32(wstr ptr) msvcr120._wfindfirst64i32
 @ cdecl _wfindnext32(long ptr) msvcr120._wfindnext32
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr) msvcr120._wfindnext32i64
 @ cdecl _wfindnext64(long ptr) msvcr120._wfindnext64
 @ cdecl _wfindnext64i32(long ptr) msvcr120._wfindnext64i32
 @ cdecl _wfopen(wstr wstr) msvcr120._wfopen
diff --git a/dlls/msvcr80/msvcr80.spec b/dlls/msvcr80/msvcr80.spec
index 12f69e7..8d682b9 100644
--- a/dlls/msvcr80/msvcr80.spec
+++ b/dlls/msvcr80/msvcr80.spec
@@ -500,11 +500,11 @@
 @ cdecl _fileno(ptr)
 @ cdecl _findclose(long)
 @ cdecl _findfirst32(str ptr)
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr)
 @ cdecl _findfirst64(str ptr)
 @ cdecl _findfirst64i32(str ptr)
 @ cdecl _findnext32(long ptr)
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr)
 @ cdecl _findnext64(long ptr)
 @ cdecl _findnext64i32(long ptr)
 @ cdecl _finite(double)
@@ -1247,11 +1247,11 @@
 @ cdecl _wexecvpe(wstr ptr ptr)
 @ cdecl _wfdopen(long wstr)
 @ cdecl _wfindfirst32(wstr ptr)
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr)
 @ cdecl _wfindfirst64(wstr ptr)
 @ cdecl _wfindfirst64i32(wstr ptr)
 @ cdecl _wfindnext32(long ptr)
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr)
 @ cdecl _wfindnext64(long ptr)
 @ cdecl _wfindnext64i32(long ptr)
 @ cdecl _wfopen(wstr wstr)
diff --git a/dlls/msvcr90/msvcr90.spec b/dlls/msvcr90/msvcr90.spec
index d137d3d..5c8e8d2 100644
--- a/dlls/msvcr90/msvcr90.spec
+++ b/dlls/msvcr90/msvcr90.spec
@@ -483,11 +483,11 @@
 @ cdecl _fileno(ptr)
 @ cdecl _findclose(long)
 @ cdecl _findfirst32(str ptr)
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr)
 @ cdecl _findfirst64(str ptr)
 @ cdecl _findfirst64i32(str ptr)
 @ cdecl _findnext32(long ptr)
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr)
 @ cdecl _findnext64(long ptr)
 @ cdecl _findnext64i32(long ptr)
 @ cdecl _finite(double)
@@ -1222,11 +1222,11 @@
 @ cdecl _wexecvpe(wstr ptr ptr)
 @ cdecl _wfdopen(long wstr)
 @ cdecl _wfindfirst32(wstr ptr)
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr)
 @ cdecl _wfindfirst64(wstr ptr)
 @ cdecl _wfindfirst64i32(wstr ptr)
 @ cdecl _wfindnext32(long ptr)
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr)
 @ cdecl _wfindnext64(long ptr)
 @ cdecl _wfindnext64i32(long ptr)
 @ cdecl _wfopen(wstr wstr)
diff --git a/dlls/msvcrt/dir.c b/dlls/msvcrt/dir.c
index a047ca4..1d91d5f 100644
--- a/dlls/msvcrt/dir.c
+++ b/dlls/msvcrt/dir.c
@@ -45,6 +45,8 @@ WINE_DEFAULT_DEBUG_CHANNEL(msvcrt);
 #undef _wfindfirst64
 #undef _wfindnext64i32
 #undef _wfindnext64
+#undef _wfindfirst32i64
+#undef _wfindnext32i64
 
 #undef _findfirst
 #undef _findfirsti64
@@ -56,6 +58,8 @@ WINE_DEFAULT_DEBUG_CHANNEL(msvcrt);
 #undef _findfirst64
 #undef _findnext64i32
 #undef _findnext64
+#undef _findfirst32i64
+#undef _findnext32i64
 
 /* INTERNAL: Translate WIN32_FIND_DATAA to finddata_t  */
 static void msvcrt_fttofd( const WIN32_FIND_DATAA *fd, struct _finddata_t* ft)
@@ -473,6 +477,30 @@ intptr_t CDECL _findfirst64(const char *fspec, struct _finddata64_t *ft)
     return ret;
 }
 
+/* INTERNAL: Translate WIN32_FIND_DATAW to wfinddata32i64_t  */
+static void msvcrt_wfttofd32i64( const WIN32_FIND_DATAW *fd, struct _wfinddata32i64_t* ft)
+{
+  DWORD dw;
+
+  if (fd->dwFileAttributes == FILE_ATTRIBUTE_NORMAL)
+    ft->attrib = 0;
+  else
+    ft->attrib = fd->dwFileAttributes;
+
+  RtlTimeToSecondsSince1970( (const LARGE_INTEGER *)&fd->ftCreationTime, &dw );
+  ft->time_create = dw;
+  RtlTimeToSecondsSince1970( (const LARGE_INTEGER *)&fd->ftLastAccessTime, &dw );
+  ft->time_access = dw;
+  RtlTimeToSecondsSince1970( (const LARGE_INTEGER *)&fd->ftLastWriteTime, &dw );
+  ft->time_write = dw;
+#ifndef SG_MUTANT_FIND32I64_LOW_SIZE
+  ft->size = ((__int64)fd->nFileSizeHigh) << 32 | fd->nFileSizeLow;
+#else
+  ft->size = fd->nFileSizeLow;
+#endif
+  wcscpy(ft->name, fd->cFileName);
+}
+
 /*********************************************************************
  *		_wfindfirst64i32 (MSVCRT.@)
  *
@@ -732,6 +760,91 @@ int CDECL _findnext64i32(intptr_t hand, struct _finddata64i32_t *ft)
     return ret;
 }
 
+/*********************************************************************
+ *		_wfindfirst32i64 (MSVCR80.@)
+ *
+ * Unicode version of _findfirst32i64.
+ */
+intptr_t CDECL _wfindfirst32i64(const wchar_t *fspec, struct _wfinddata32i64_t *ft)
+{
+  WIN32_FIND_DATAW find_data;
+  HANDLE hfind;
+
+  hfind  = FindFirstFileW(fspec, &find_data);
+  if (hfind == INVALID_HANDLE_VALUE)
+  {
+    msvcrt_set_errno(GetLastError());
+    return -1;
+  }
+  msvcrt_wfttofd32i64(&find_data, ft);
+  TRACE(":got handle %p\n", hfind);
+  return (intptr_t)hfind;
+}
+
+static int finddata32i64_wtoa(const struct _wfinddata32i64_t *wfd, struct _finddata32i64_t *fd)
+{
+    fd->attrib = wfd->attrib;
+    fd->time_create = wfd->time_create;
+    fd->time_access = wfd->time_access;
+    fd->time_write = wfd->time_write;
+    fd->size = wfd->size;
+    return convert_wcs_to_acp_utf8(wfd->name, fd->name, ARRAY_SIZE(fd->name));
+}
+
+/*********************************************************************
+ *		_findfirst32i64 (MSVCR80.@)
+ *
+ * 32-bit time, 64-bit size version of _findfirst.
+ */
+intptr_t CDECL _findfirst32i64(const char *fspec, struct _finddata32i64_t *ft)
+{
+    struct _wfinddata32i64_t wft;
+    wchar_t *fspecW = NULL;
+    intptr_t ret;
+
+    if (fspec && !(fspecW = wstrdupa_utf8(fspec))) return -1;
+    ret = _wfindfirst32i64(fspecW, &wft);
+    free(fspecW);
+    if (ret != -1 && !finddata32i64_wtoa(&wft, ft))
+    {
+        _findclose(ret);
+        return -1;
+    }
+    return ret;
+}
+
+/*********************************************************************
+ *		_wfindnext32i64 (MSVCR80.@)
+ *
+ * Unicode version of _findnext32i64.
+ */
+int CDECL _wfindnext32i64(intptr_t hand, struct _wfinddata32i64_t *ft)
+{
+  WIN32_FIND_DATAW find_data;
+
+  if (!FindNextFileW((HANDLE)hand, &find_data))
+  {
+    *_errno() = ENOENT;
+    return -1;
+  }
+
+  msvcrt_wfttofd32i64(&find_data, ft);
+  return 0;
+}
+
+/*********************************************************************
+ *		_findnext32i64 (MSVCR80.@)
+ */
+int CDECL _findnext32i64(intptr_t hand, struct _finddata32i64_t *ft)
+{
+    struct _wfinddata32i64_t wft;
+    int ret;
+
+    ret = _wfindnext32i64(hand, &wft);
+    if (!ret && !finddata32i64_wtoa(&wft, ft)) ret = -1;
+    return ret;
+}
+
 /*********************************************************************
  *		_getcwd (MSVCRT.@)
  *
diff --git a/dlls/ucrtbase/ucrtbase.spec b/dlls/ucrtbase/ucrtbase.spec
index aaee003..b6c0c5e 100644
--- a/dlls/ucrtbase/ucrtbase.spec
+++ b/dlls/ucrtbase/ucrtbase.spec
@@ -316,11 +316,11 @@
 @ cdecl _fileno(ptr)
 @ cdecl _findclose(long)
 @ cdecl _findfirst32(str ptr)
-@ stub _findfirst32i64
+@ cdecl _findfirst32i64(str ptr)
 @ cdecl _findfirst64(str ptr)
 @ cdecl _findfirst64i32(str ptr)
 @ cdecl _findnext32(long ptr)
-@ stub _findnext32i64
+@ cdecl _findnext32i64(long ptr)
 @ cdecl _findnext64(long ptr)
 @ cdecl _findnext64i32(long ptr)
 @ cdecl _finite(double)
@@ -940,11 +940,11 @@
 @ cdecl _o__fileno(ptr) _fileno
 @ cdecl _o__findclose(long) _findclose
 @ cdecl _o__findfirst32(str ptr) _findfirst32
-@ stub _o__findfirst32i64
+@ cdecl _o__findfirst32i64(str ptr) _findfirst32i64
 @ cdecl _o__findfirst64(str ptr) _findfirst64
 @ cdecl _o__findfirst64i32(str ptr) _findfirst64i32
 @ cdecl _o__findnext32(long ptr) _findnext32
-@ stub _o__findnext32i64
+@ cdecl _o__findnext32i64(long ptr) _findnext32i64
 @ cdecl _o__findnext64(long ptr) _findnext64
 @ cdecl _o__findnext64i32(long ptr) _findnext64i32
 @ cdecl _o__flushall() _flushall
@@ -1492,11 +1492,11 @@
 @ cdecl _o__wexecvpe(wstr ptr ptr) _wexecvpe
 @ cdecl _o__wfdopen(long wstr) _wfdopen
 @ cdecl _o__wfindfirst32(wstr ptr) _wfindfirst32
-@ stub _o__wfindfirst32i64
+@ cdecl _o__wfindfirst32i64(wstr ptr) _wfindfirst32i64
 @ cdecl _o__wfindfirst64(wstr ptr) _wfindfirst64
 @ cdecl _o__wfindfirst64i32(wstr ptr) _wfindfirst64i32
 @ cdecl _o__wfindnext32(long ptr) _wfindnext32
-@ stub _o__wfindnext32i64
+@ cdecl _o__wfindnext32i64(long ptr) _wfindnext32i64
 @ cdecl _o__wfindnext64(long ptr) _wfindnext64
 @ cdecl _o__wfindnext64i32(long ptr) _wfindnext64i32
 @ cdecl _o__wfopen(wstr wstr) _wfopen
@@ -2077,11 +2077,11 @@
 @ cdecl _wexecvpe(wstr ptr ptr)
 @ cdecl _wfdopen(long wstr)
 @ cdecl _wfindfirst32(wstr ptr)
-@ stub _wfindfirst32i64
+@ cdecl _wfindfirst32i64(wstr ptr)
 @ cdecl _wfindfirst64(wstr ptr)
 @ cdecl _wfindfirst64i32(wstr ptr)
 @ cdecl _wfindnext32(long ptr)
-@ stub _wfindnext32i64
+@ cdecl _wfindnext32i64(long ptr)
 @ cdecl _wfindnext64(long ptr)
 @ cdecl _wfindnext64i32(long ptr)
 @ cdecl _wfopen(wstr wstr)
